bot/exts/info/tags.py

```diff
@@ -24,145 +29,229 @@
 FOOTER_TEXT = f"To show a tag, type {constants.Bot.prefix}tags <tagname>."
 
 
+class COOLDOWN(enum.Enum):
+    """Sentinel value to signal that a tag is on cooldown."""
+
+    obj = object()
+
+
+class TagIdentifier(NamedTuple):
+    """Stores the group and name used as an identifier for a tag."""
+
+    group: Optional[str]
+    name: str
+
+    def get_fuzzy_score(self, fuzz_tag_identifier: TagIdentifier) -> float:
+        """Get fuzzy score, using `fuzz_tag_identifier` as the identifier to fuzzy match with."""
+        if self.group is None:
+            if fuzz_tag_identifier.group is None:
+                # We're only fuzzy matching the name
+                group_score = 1
+            else:
+                # Ignore tags without groups if the identifier contains a group
+                return .0
+        else:
+            if fuzz_tag_identifier.group is None:
+                # Ignore tags with groups if the identifier does not have a group
+                return .0
+            else:
+                group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
```

Comment:
I was able to simplify this code a bit considering there's 3 cases:
- One has a group and the other doesn't (XOR)
- Identical groups (potentially both None)
- Different strings and should be fuzzy searched

```suggestion
        if self.group is None != fuzzy_tag_identifier.group is None:
            # Ignore tags without groups if the identifier has a group and vice versa
            return .0

        if self.group == fuzzy_tag_identifier.group:
            # Completely identical, or both None
            group_score = 1
        else:
            group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
```

Reply:
The  `self.group is None != fuzzy_tag_identifier.group is None:` comparison won't work. we'd need to do something like (not sure if there's any nicer way to check that either of  them is None and they aren't equal)
```suggestion
        if (self.group is None or fuzz_tag_identifier.group is None) and self.group != fuzz_tag_identifier.group:
            # Ignore tags without groups if the identifier has a group and vice versa.
            return .0
        if self.group == fuzz_tag_identifier.group:
            # Completely identical, or both None.
            group_score = 1
        else:
            group_score = _fuzzy_search(fuzz_tag_identifier.group, self.group)
```
I think that's a bit easier to read overall but the first condition could be confusing at a glance

Reply:
No `self.group is None != fuzzy_tag_identifier.group is None:` will work, because they become bools. Bools only have 2 states, True or False. So we have 4 cases:
- both `self.group` and `fuzzy_tag_identifier.group` are None (`True != True` *False*)
- `self.group` is None and `fuzzy_tag_identifer.group` is set (`True != False` *True*)
- `self.group` is set and `fuzzy_tag_identifier.group` is None (`False != True` *True*)
- both `self.group` and `fuzzy_tag_identifier.group` are set (`False != False` *False*)

We want to return when it is True, which is exaclty what is happening here.

Reply:
With how it's chained, it impossible for the condition to be True `self.group is None != fuzzy_tag_identifier.group is None` creates a `!=` comparison between `None` and `fuzzy_tag_identifier.group` which is then compared to be equal to `None`. The long equivalent (and how python handles it) would be 
`self_group is None and None != fuzzy_tag_identifier_group and fuzzy_tag_identifier_group is None`

```py
In [149]: for self_group, fuzzy_tag_identifier_group in itertools.product((None, "ba"), repeat=2):
     ...:     print(self_group, fuzzy_tag_identifier_group)
     ...:     print(self_group is None != fuzzy_tag_identifier_group is None)
     ...:     print((self_group is None or fuzzy_tag_identifier_group is None) and self_group != fuzzy_tag_identifier_g
     ...: roup)
     ...:     print()
     ...:
None None
False
False

None ba
False
True

ba None
False
True

ba ba
False
False
```

Reply:
Oh I see, so just add the parenthesis: `(self.group is None) != (fuzzy_tag_identifier.group is None)`

Reply:
Ah, yes that should work, sorry I couldn't make sense of the comparison before